Convert types to/from database using Postgres

timheap opened the pull request:
Instead of relying on Django and its `get_db_value`/`from_db_value`, the type is registered more deeply with Postgres. This allows the type to be used in ArrayFields and other places where typecasting is required.

danni commented on postgres_composite_types/__init__.py:
```diff
@@ -317,6 +290,23 @@ def __eq__(self, other):
+        # For some reason, adapt().getquoted() returns bytes, but AsIs requires
+        # a str. Using bytes in AsIs leads to a bad time. Formatting a series
+        # of bytes is also not fun
+        query = b''.join([
```
Can you not use a format string + encode/decode here?

danni replied:
Maybe a format string.

danni replied:
I'm curious about tests for storing utf-8 in the database, this generally works. Also a test for the `ArrayField`.

timheap commented:
OK, everything has been rewritten. Same basic premise, but a new approach that uses the psycopg internals correctly. I've added a bunch of comments, so hopefully you do not have to do as much spelunking through psycopg code as I did to understand it!

I've also added tests for more complicated use-cases, which was the inspiration for this in the first place. ArrayFields with CompositeTypes, and nested CompositeTypes are covered.

danni commented on postgres_composite_types/__init__.py:
```diff
@@ -43,71 +43,76 @@
+    def __conform__(self, protocol):
+        # Required for nested composite types
+        if protocol is ISQLQuote:
+            return self
```
What happens otherwise?

danni replied:
Generally some comments explaining where you can see how this implementation works would be useful for the next hapless reader trying to understand how the frack this works.

danni commented on postgres_composite_types/__init__.py:
```diff
@@ -43,71 +43,76 @@
+        """
+        if self.prepared_value is None:
+            raise RuntimeError(
+                "%(name)s.prepare() must be called before %(name)s.getquoted()"
```
Prefer `.format()`

danni commented:
Add yourself to the README as an author, you've basically rewritten half of the code at this point :)

timheap commented:
Some more commits pushed
